File: solution/c#/Day20/Day20/Domain/Yahtzee/Constrain.Input/Roll.cs
using System.Collections.Immutable;
using LanguageExt;

namespace Day20.Domain.Yahtzee.Constrain.Input
{
    public sealed class Roll : IEquatable<Roll>
    {
        private const int RollLength = 5;
        private const int MinimumDie = 1;
        private const int MaximumDie = 6;
        public int[] Dice { get; }
        private Roll(int[] dice) => Dice = dice;

        public static Either<ParsingError, Roll> Parse(int[] dice)
        {
            if (HasInvalidLength(dice))
            {
                return new ParsingError("Invalid dice... A roll should contain 5 dice.");
            }

            return ContainsInvalidDie(dice)
                ? new ParsingError("Invalid die value. Each die must be between 1 and 6.")
                : new Roll(dice.ToImmutableArray().ToArray());
        }

        public Dictionary<int, int> GroupDieByFrequency()
            => Dice.GroupBy(x => x).ToDictionary(g => g.Key, g => g.Count());

        private static bool HasInvalidLength(int[] dice) => dice is not {Length: RollLength};
        private static bool ContainsInvalidDie(IEnumerable<int> dice) => !dice.All(IsValidDie);
        private static bool IsValidDie(int die) => die is >= MinimumDie and <= MaximumDie;
        public override string ToString() => Dice.ToString()!;

        #region IEquatable

        public override int GetHashCode() => Dice.GetHashCode();
        public static bool operator ==(Roll roll, Roll other) => roll.Equals(other);
        public static bool operator !=(Roll roll, Roll other) => !(roll == other);
        public bool Equals(Roll? other) => Dice.Equals(other!.Dice);
        public override bool Equals(object? obj) => obj is Roll && Equals(Dice);

        #endregion
    }

    public sealed record ParsingError(string Reason);

    public static class Extensions
    {
        public static Either<ParsingError, Roll> ToRoll(this int[] dice) => Roll.Parse(dice);
    }
}

File: solution/ts/day20/src/domain/yahtzee/constrain-input/roll.ts
import {Either, left, right} from "fp-ts/Either";

export type ParsingError = { message: string }

export class Roll {
    private static readonly ROLL_LENGTH: number = 5;
    private static readonly MINIMUM_DIE: number = 1;
    private static readonly MAXIMUM_DIE: number = 6;

    private constructor(readonly dice: number[]) {
    }

    public static from(dice: number[]): Either<ParsingError, Roll> {
        if (this.hasInvalidLength(dice)) return left({message: "Invalid dice... A roll should contain 5 dice."});
        return this.containsInvalidDie(dice)
            ? left({message: "Invalid die value. Each die must be between 1 and 6."})
            : right(new Roll(dice));
    }

    private static hasInvalidLength = (dice: number[]): boolean =>
        !dice || dice.length !== this.ROLL_LENGTH;

    private static containsInvalidDie = (dice: number[]): boolean =>
        dice.some(die => this.isInvalidDie(die));

    private static isInvalidDie = (die: number): boolean =>
        die < this.MINIMUM_DIE || die > this.MAXIMUM_DIE;

    public groupDieByFrequency = (): Record<number, number> =>
        this.dice.reduce((freq, die) => {
            freq[die] = (freq[die] || 0) + 1;
            return freq;
        }, {} as Record<number, number>);

    public toString = (): string => `[${this.dice.sort().join(", ")}]`;
}